splink/blocking_rule_library.py

```diff
@@ -0,0 +1,169 @@
+                    sql_condition = _translate_sql_string(
+                        sql_condition,
+                        sql_dialect.sqlglot_name,
+                        base_dialect_sqlglot_name,
+                    )
+                # if we hit a sqlglot error, assume users knows what they are doing,
+                # e.g. it is something custom / unknown to sqlglot
+                # error will just appear when they try to use it
+                except TokenError:
+                    pass
+        return sql_condition
+
+
+class _Merge(BlockingRuleCreator):
+    @final
+    def __init__(
+        self,
+        *blocking_rules: Union[BlockingRuleCreator, dict],
+        salting_partitions=None,
+        arrays_to_explode=None,
+    ):
+        super().__init__(
+            salting_partitions=salting_partitions, arrays_to_explode=arrays_to_explode
+        )
+        num_levels = len(blocking_rules)
+        if num_levels == 0:
+            raise ValueError(
+                f"Must provide at least one blocking rule to {type(self)}()"
+            )
+        self.blocking_rules = blocking_rules
+
+    @property
+    def salting_partitions(self):
+        if (
+            hasattr(self, "_salting_partitions")
+            and self._salting_partitions is not None
+        ):
+            return self._salting_partitions
+
+        return max(
+            [
+                br.salting_partitions
+                for br in self.blocking_rules
+                if br.salting_partitions is not None
+            ],
+            default=None,
+        )
+
+    @property
+    def arrays_to_explode(self):
+        if hasattr(self, "_arrays_to_explode"):
+            return self._arrays_to_explode
+
+        if any([br.arrays_to_explode for br in self.blocking_rules]):
+            raise ValueError("Cannot merge blocking rules with arrays_to_explode")
+        return None
+
+    @final
+    def create_sql(self, sql_dialect: SplinkDialect) -> str:
+        return f" {self._clause} ".join(
+            map(lambda cl: f"({cl.create_sql(sql_dialect)})", self.blocking_rules)
+        )
+
+
+class And(_Merge):
+    _clause = "AND"
+
+
+class Or(_Merge):
+    _clause = "OR"
+
+
+class Not(BlockingRuleCreator):
+    def __init__(self, blocking_rule_creator):
+        self.blocking_rule_creator = blocking_rule_creator
+
+    @property
+    def salting_partitions(self):
+        return self.blocking_rule_creator.salting_partitions
+
+    @property
+    def arrays_to_explode(self):
+        if self.blocking_rule_creator.arrays_to_explode:
+            raise ValueError("Cannot use arrays_to_explode with Not")
+        return None
+
+    @final
+    def create_sql(self, sql_dialect: SplinkDialect) -> str:
+        return f"NOT ({self.blocking_rule_creator.create_sql(sql_dialect)})"
+
+
+def block_on(
+    *col_names_or_exprs: Union[str, ColumnExpression],
+    salting_partitions=None,
+    arrays_to_explode=None,
+) -> BlockingRuleCreator:
```

Comment:
I wonder if it might be helpful to give an error here if people use the old syntax, given that this is possibly a more subtle change compared to other API tweaks?
For instance currently if you use `block_on(["first_name", "surname"])` you will hit `AttributeError: 'NoneType' object has no attribute 'sql_dialect'` at some point once the linker is doing its thing, so it might be handy to nudge users in the right direction rather than them trying to unpick that

Reply:
Good point - done